dashboard/app/controllers/level_starter_assets_controller.rb

```diff
@@ -0,0 +1,66 @@
+class LevelStarterAssetsController < ApplicationController
+  before_action :set_level
+
+  S3_BUCKET = 'cdo-v3-assets'.freeze
+  S3_PREFIX = 'starter_assets/'.freeze
+
+  # GET /level_starter_assets/:level_name
+  def show
+    starter_assets = @level.starter_assets.map do |friendly_name, guid_name|
+      file_obj = get_file_object(prefix(guid_name))
+
+      if file_obj.size.zero?
+        nil
+      else
+        {
+          filename: friendly_name,
+          category: file_mime_type(File.extname(guid_name)),
+          size: file_obj.size,
+          timestamp: file_obj.last_modified
+        }
+      end
+    end.compact
+
+    render json: {starter_assets: starter_assets}
+  end
+
+  # GET /level_starter_assets/:level_name/:filename
+  # Returns requested file body as an IO stream.
+  def file
+    friendly_name = "#{params[:filename]}.#{params[:format]}"
+    guid_name = @level.starter_assets[friendly_name]
+    file_obj = get_file_object(prefix(guid_name))
+    content_type = file_content_type(File.extname(guid_name))
+    send_data read_file(file_obj), type: content_type, disposition: 'inline'
```

Comment:
Does this download the entire file into memory on our server first before streaming it to the client? Is there a way to start streaming the content immediately instead?

Reply:
I think the answer to your first question is yes, but here's more detail since I'm not positive:

The `read_file` is a helper that maps to `file_obj.get.body.read` -- `body` is returned from the S3 SDK as an [IO](https://ruby-doc.org/core-2.5.1/IO.html), then I'm using Ruby's [`IO#read`](https://ruby-doc.org/core-2.5.1/IO.html#method-i-read) method to extract the file data and Rails' [`send_data`](https://apidock.com/rails/ActionController/Streaming/send_data) method to send the file to the client.

Levelbuilders currently use starter assets by uploading an image through our FilesApi, then the level renders that image using the [`/v3/assets/:channel_id/:filename`](https://github.com/code-dot-org/code-dot-org/blob/161ab002042bd31600d2045678f1147f910c067d/shared/middleware/files_api.rb#L137-L144) endpoint, so this new endpoint is meant to mimic that FilesApi endpoint.

Reply:
Okay, it looks like that `IO#read` is reading the entire contents into a String, so I think the answer is yes.

Instead of passing the result of `file_obj.get.body.read` to `send_data`, does it work if you pass `file_obj.get.body` itself? I think that would avoid that extra step.

Reply:
No, returning `file_obj.get.body` doesn't work, but I did mess around with using [`read_nonblock`](https://ruby-doc.org/core-2.6.3/IO.html#method-i-read_nonblock) instead of `read` and it seems to work:

```
file_obj.get.body.read_nonblock(file_obj.size)
```

but I don't feel like I really understand what read_nonblock does

Reply:
Regarding performance, we should be able to set a very long cache lifetime when starter assets are requested for read, which should get us most of the way to good performance.

Reply:
thanks, dave! winter suggested the same, so https://github.com/code-dot-org/code-dot-org/pull/30218/commits/16b3f4968ec77d28267d264df9cdc8dd05d34502 adds a 1-hour cache expiration to the response using [expires_in](https://apidock.com/rails/ActionController/ConditionalGet/expires_in), which sets the Cache-Control header. 

is this enough to cache the response in Cloudfront? is there a better/standardized way to do this?

also, i grabbed some response headers if folks are interested--

response headers from initial request to `/level_starter_assets/:level_name/:filename` with Cache-Control being set:
```
HTTP/1.1 200 OK
Content-Type: image/png
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Date: Mon, 19 Aug 2019 21:15:32 GMT
Content-Disposition: inline
Content-Transfer-Encoding: binary
Cache-Control: max-age=3600, public
ETag: W/"cc9dc8917ec61f2a02815a6cb26f4fae"
Content-Length: 14699
Vary: Host
X-Content-Digest: 537722607b2294661f42b967a1bb64e2a4299822
Age: 1
X-Rack-Cache: miss, store
X-Request-Id: 59b4b502-410e-4d16-9157-499cc7de00a1
X-Runtime: 0.454960
Connection: keep-alive
Server: thin
```

response headers from a subsequent request:
```
HTTP/1.1 200 OK
Content-Type: image/png
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Date: Mon, 19 Aug 2019 21:15:32 GMT
Content-Disposition: inline
Content-Transfer-Encoding: binary
Cache-Control: max-age=3600, public
ETag: W/"cc9dc8917ec61f2a02815a6cb26f4fae"
Content-Length: 14699
Vary: Host
X-Content-Digest: 537722607b2294661f42b967a1bb64e2a4299822
Age: 105
X-Rack-Cache: fresh
X-Request-Id: f7d97d50-fc79-4b61-b1a0-733961f484d0
X-Runtime: 0.124507
Connection: keep-alive
Server: thin
```

Reply:
I'm not 100% sure this applies here, but just wanted to share a pointer to a different place where we have to read the entire file rather than streaming it to the client: https://github.com/code-dot-org/code-dot-org/blob/a26a95f71fc20f83fb0b24bd4b9629d550bb9016/shared/middleware/files_api.rb#L405-L409

this might not be relevant here yet since we're not conditionally rejecting the request, also this is rails and not just sinatra etc. but wanted to include the link in case the problem described there comes up here.

Reply:
`expires_in` looks right, and the cache headers look good to me.

there may be one more step to make this route cacheable. you need to tell varnish / cloudfront to ignore cookies and headers, so that it doesn't try to cache a separate copy for every user, language, etc. here is an example of how to do that: https://github.com/code-dot-org/code-dot-org/blob/d16fc9addb88600299d92889c29a019e722f5089/cookbooks/cdo-varnish/libraries/http_cache.rb#L222-L229

I'm not really sure how it works when you want the same route to be non-cacheable for PUT or POST requests but be cacheable for GET requests. that's why we used a different route like `/v3/files-public` instead of `/v3/files` at the above link. so one idea would be to split the API so that the publicly cacheable route has a different prefix.

TBH I think this caching change is getting big enough that it might be better to split it out into another PR. it's fine to leave `expires_in` how you have it but just keep in mind that you might have more work to do before it takes effect.

Reply:
ohh, interesting -- thank you! yeah, that sounds like it should belong in its own PR. winter and i also discussed refactoring the client so that it has a reference to the assets' S3 URLs so that we can hit S3 directly, rather than having separate endpoints on our servers that just serve content from S3. so i think i will size that work up and probably go that route, since it will have the same benefit and will also remove complexity from this feature